src/command.test.js

```diff
@@ -1,6 +1,144 @@
-const { describe, it, expect } = require('@jest/globals')
-describe('command()', () => {
-  it('happy path', () => {
-    expect(true).toBe(true)
+const { describe, it, expect, beforeEach } = require('@jest/globals')
+const { Command } = require('./command')
+
+describe('Command', () => {
+  describe('empty command', () => {
+    class TestCommand extends Command {
+      schema =
+        {
+
+        }
+
+      execute () {
+        return 'Result!'
+      }
+    }
+
+    describe('run', () => {
+      it('Is successful', async () => {
+        const outcome = await TestCommand.run()
+        expect(outcome.success).toBe(true)
+        expect(outcome.result).toBe('Result!')
+      })
+    })
+  })
+
+  describe('command with inputs', () => {
+    class TestCommand extends Command {
+      schema =
+        {
+          input1: { type: 'string', required: true },
+          input2: { type: 'string', default: 'default input2' }
+        }
+
+      execute () {
+        return this.inputs.input1
+      }
+
+      validate () {
+        if (this.inputs.input1 === 'invalid input1') {
+          this.addInputError('input1', 'invalid', 'invalid value!')
+        }
+      }
+    }
+
+    describe('run', () => {
+      it('Is successful', async () => {
+        const outcome = await TestCommand.run({ input1: 'This just works' })
+        expect(outcome.success).toBe(true)
+        expect(outcome.result).toBe('This just works')
+      })
+
+      describe('required input missing', () => {
+        it('Is not successful', async () => {
+          const outcome = await TestCommand.run()
+          expect(outcome.success).toBe(false)
+          expect(outcome.result).toBeNull()
+          expect(outcome.errors.input1[0][0]).toEqual('missing')
+        })
+      })
+      describe('manually adding an input error', () => {
+        it('Is not successful', async () => {
+          const outcome = await TestCommand.run({ input1: 'invalid input1' })
+          expect(outcome.success).toBe(false)
+          expect(outcome.result).toBeNull()
+          expect(outcome.errors.input1[0][0]).toEqual('invalid')
+        })
+      })
+      describe('runtime error', () => {
+        class AnotherTestCommand extends TestCommand {
+          execute () {
+            this.operation1()
+            this.operation2()
+            this.operation3()
+          }
+
+          operation1 () {
+            return 'operation 1 ok'
+          }
+
+          operation2 () {
+            this.addRuntimeError('operation 2 exploded!')
+          }
+
+          operation3 () {
+            return 'operation 3 ok'
+          }
+        }
+
+        beforeEach(() => {
+          const { jest } = require('@jest/globals')
+          jest.spyOn(AnotherTestCommand.prototype, 'operation1')
+          jest.spyOn(AnotherTestCommand.prototype, 'operation2')
+          jest.spyOn(AnotherTestCommand.prototype, 'operation3')
```

Comment:
these require a call to `jest.restoreAllMocks` to avoid leaking these into other tests I think. You could grep labs for usages.

Reply:
Addressed!